### PR TITLE
Added MultilayerPerceptron

### PR DESCRIPTION
**Pull request by Abi1024**
Multilayer Neural Network implementation for Saul, with support for sparse feature vectors, various activation functions and L-2 regularization.

### CODE REVIEW

**Comment by kordjamshidi** [comment]
Thank you very much @Abi1024 . @bhargav @danyaljj please review at your earliest convenience.

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+        return  Math.log(1.0d + Math.exp(x));
+    }
+
+    //softplus function derivative
+    private double softplus_derivative(double x) {
+        return sigmoid(x);
+    }
+
+    public ScoreSet scores(Object example){
+        return null;
```
Comment (kordjamshidi) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: We need scores here but this returns null, right?

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+        }
+    }
+
+    //Construct a perceptron with the default parameters defined above.
+    public MultilayerPerceptron(){
+        this(default_num_inputs,default_num_outputs,default_hidden_layers,default_activation_function);
+    }
+
+    //Construct a perceptron with user specified parameters.
+    public MultilayerPerceptron(int num_inputs, int num_outputs, int[] hidden_layers, int [] activation_function){
```
Comment (danyaljj) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: Could you clarify each input value with a javadoc description?

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    //Default learning parameters
+    private static double default_learning_rate = 0.1d;
+    private static double default_momentum = 0.1d;
+    private static double default_lambda = 0.00001d;
+
+    //Default network parameters. For sparse feature vectors, the network will automatically resize itself.
+    private static int default_num_inputs = 128;
+    private static int[] default_hidden_layers =  new int[]{};
+    private static int[] default_activation_function = {SIGMOID};
+    private static int default_num_outputs = 2;
```
Comment (danyaljj) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: Does "2" here mean "one output variables for 2 classes"?

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java (hunk 1 above)
Comment (danyaljj) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: If I understand correctly we have to call the `infer` function here?

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+    public ScoreSet scores(Object example){
+        return null;
+    }
+
+    public ScoreSet scores(int[] exampleFeatures, double[] exampleValues) {
+        return null;
+    }
+
+    public void write(ExceptionlessOutputStream out) {
+        out.close();
```
Comment (danyaljj) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: Can you add proper TODO message here, so that we remember to implement it? 
To be honest classifiers are not that useful in Saul, if their serialization/deserialization is not working. 

FYI @kordjamshidi

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+        compute_activation(input);
+        output_errors(output);
+        compute_gradients(input);
+        update_weights(learning_rate,lambda, momentum);
+    }
+
+    /**
+     * Trains the neural network on a sparse input vector, using default learning parameters.
+     *
+     * @TODO Only works for binary classification now.
```
Comment (danyaljj) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: could you clarify this? Do you mean, the output vector can only be a vector of 1.0/0.0 ?

**Comment by danyaljj** [comment]
Looks (almost) good to me. Passing it to @bhargav for a final look.

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+        private double[][] gradients;
+        private double[][] prev_gradients;
+        private double[] pre_activation;
+        private double[] post_activation;
+        private double[] errors;
+        private int activation_function;
+
+        // Initializes a layer of the neural network. Weights are all initialized to 0.
+        private Layer(int num_inputs, int num_outputs, int activation_function){
+            weights = new double[num_inputs+1][num_outputs];
```
Comment (bhargav) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: Minor: I believe primitives in Java are initialized to 0 by default. You might not need to fill the arrays with 0.

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+        int best_index = -1;
+        double best_value = -1;
+        for (int i = 0; i < layers[layers.length-1].errors.length; i++){
+            if (layers[layers.length-1].post_activation[i] > best_value){
+                best_value = layers[layers.length-1].post_activation[i];
+                best_index = i;
+            }
+        }
+        if (best_index == 0){ //for binary classifiers used in Twitter sentiment analysis
+            return new DiscretePrimitiveStringFeature(containingPackage, name, "",
```
Comment (bhargav) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: The returned features here would be problematic for other tasks. The labels might not be "negative"/"positive"

Comment (bhargav) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: You should use something to the effect of `labelLexicon.lookupKey(i)` on the best_index.

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+            pre_activation = new double[num_outputs];
+            Arrays.fill(pre_activation,0.0d);
+            post_activation = new double[num_outputs];
+            Arrays.fill(post_activation,0.0d);
+            errors = new double[num_outputs];
+            Arrays.fill(errors,0.0d);
+            this.activation_function = activation_function;
+        }
+
+        public void write(ExceptionlessOutputStream out){
```
Comment (bhargav) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: Quick Question: while serializing models, we should not save the gradients/prev_gradients and activations right? We will have to evaluate them on examples in the forward(-backwards) pass when we load a model to test or train.

File: saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+        Arrays.fill(output,0.0d);
+        output[exampleLabels[0]] = 1.0d;
+        int maximum_index = -1;
+        for (int i = 0; i < exampleFeatures.length; i++){
+            if  (exampleFeatures[i] > maximum_index){
+                maximum_index = exampleFeatures[i];
+            }
+        }
+        if (layers[0].weights.length < maximum_index + 2 )
+            resize_inputs(maximum_index + 10000);
```
Comment (bhargav) [inline saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java]: For later: It might be better to do a pre-extract step and calculate the final label lexicon size. This might add a high computation overhead if we do not have 10k more features.

**Comment by Abi1024** [comment]
Hi, I just incorporated all of @bhargav 's suggestions and also added regression for the facebook circumplex dataset. If someone could take a look sometime, that would be much appreciated. Thanks!

**Comment by kordjamshidi** [comment]
@bhargav please see if this is ok with the recent changes. However, I am not sure if we want to merge the new app of circumplex (sentiment analysis) model also into Saul or not. @Abi1024 can remove that application, maybe, and just keep the implementation of the Learners. Let him know if removing the App is preferred @danyaljj @bhargav .

**Comment by kordjamshidi** [comment]
Oh, I see @bhargav has approved the changes already. So we merge the app too? Any opinion, comments?

**Comment by danyaljj** [comment]
Agreed. Unless the application is complete  (fully functional, with readme etc), I would rather not checking it in.

**Comment by kordjamshidi** [comment]
Thanks @Abi1024 for removing the app, I m going to merge then.
